app/models/payment_code.rb

```
@@ -0,0 +1,31 @@
+class PaymentCode < ApplicationRecord
+  CONFUSED_CHARS = %w(B 8 G 6 I 1 O 0 S 5 Z 2)
+
+  validates_presence_of :code
+  validates_uniqueness_of :code
+
+  before_validation :generate_code, on: :create
+  after_rollback :regenerate_code
+
+  belongs_to :student, subsystem: :course_membership, inverse_of: :payment_codes, optional: true
+
+  attr_accessor :prefix
+
+  def generate_code
+    unless prefix.present?
+      errors.add :prefix, :blank
+      throw :abort
+    end
+
+    base = [*'0'..'9', *'A'..'Z'] - CONFUSED_CHARS
+    post = Array.new(10) { base.sample }.join
+    self.code = "#{prefix.parameterize.upcase}-#{post}"
+  end
+
+  def regenerate_code
+    if errors.types[:code]&.include?(:taken)
+      generate_code
+      save
+    end
+  end
```

this method makes me nervous.  A code should never change, because once we hand them out we MUST honor them.  It'd suck if a student somehow paid for a code that we'd overwritten and no longer had a record of.  

I see the code column has a unique index, and since collisions should be very unlikely, how about adding a try/catch block in the generate codes routine and just loop through again if it does happen

---

I also think @Dantemss has an `immutablerecord`  class somewhere, we should use that so that codes cannot be deleted

---

@nathanstitt Thoughts if https://github.com/openstax/tutor-server/pull/2333/commits/e4a4a1570e074db74e397b89150ba9caad2c7e22 addresses those concerns? If not I'm totally good with moving this behavior into the routine 👍

---

think that's good but how about having it throw an exception if it's persisted?  Right now it'll silently fail to set_code so whatever's calling it wouldn't know that it shouldn't be doing so